From a467ee2002f398696183fa821d159ed3219e9d63 Mon Sep 17 00:00:00 2001
From: Aditya Garg <85610623+AdityaGarg8@users.noreply.github.com>
Date: Sat, 19 Mar 2022 14:31:18 +0530
Subject: [PATCH] Force enable intel iommu on T2 Macs

---
 drivers/iommu/intel/iommu.c | 36 ++++++++++++++++++++++++++++++++++++
 1 file changed, 36 insertions(+)

diff --git a/drivers/iommu/intel/iommu.c b/drivers/iommu/intel/iommu.c
index 5b196cfe9..c1e8450b4 100644
--- a/drivers/iommu/intel/iommu.c
+++ b/drivers/iommu/intel/iommu.c
@@ -24,6 +24,7 @@
 #include <linux/pci.h>
 #include <linux/dmar.h>
 #include <linux/dma-map-ops.h>
+#include <linux/dmi.h>
 #include <linux/mempool.h>
 #include <linux/memory.h>
 #include <linux/cpu.h>
@@ -86,6 +87,13 @@
 #define LEVEL_STRIDE		(9)
 #define LEVEL_MASK		(((u64)1 << LEVEL_STRIDE) - 1)
 
+/* Force enable IOMMU on some computers */
+#define FORCE_ENABLE_IOMMU(vendor, product) \
+		 .matches = { \
+			DMI_MATCH(DMI_BOARD_VENDOR, vendor), \
+			DMI_MATCH(DMI_PRODUCT_NAME, product), \
+		},
+
 static inline int agaw_to_level(int agaw)
 {
 	return agaw + 2;
@@ -405,8 +413,36 @@ static void init_translation_status(struct intel_iommu *iommu)
 		iommu->flags |= VTD_FLAG_TRANS_PRE_ENABLED;
 }
 
+static const struct dmi_system_id force_iommu_quirk[] = {
+
+	{ FORCE_ENABLE_IOMMU("Apple Inc.", "MacBookPro15,1") },
+	{ FORCE_ENABLE_IOMMU("Apple Inc.", "MacBookPro15,2") },
+	{ FORCE_ENABLE_IOMMU("Apple Inc.", "MacBookPro15,3") },
+	{ FORCE_ENABLE_IOMMU("Apple Inc.", "MacBookPro15,4") },
+	{ FORCE_ENABLE_IOMMU("Apple Inc.", "MacBookPro16,1") },
+	{ FORCE_ENABLE_IOMMU("Apple Inc.", "MacBookPro16,2") },
+	{ FORCE_ENABLE_IOMMU("Apple Inc.", "MacBookPro16,3") },
+	{ FORCE_ENABLE_IOMMU("Apple Inc.", "MacBookPro16,4") },
+	{ FORCE_ENABLE_IOMMU("Apple Inc.", "MacBookAir8,1") },
+	{ FORCE_ENABLE_IOMMU("Apple Inc.", "MacBookAir8,2") },
+	{ FORCE_ENABLE_IOMMU("Apple Inc.", "MacBookAir9,1") },
+	{ FORCE_ENABLE_IOMMU("Apple Inc.", "MacMini8,1") },
+	{ FORCE_ENABLE_IOMMU("Apple Inc.", "MacPro7,1") },
+	{ FORCE_ENABLE_IOMMU("Apple Inc.", "iMac20,1") },
+	{ FORCE_ENABLE_IOMMU("Apple Inc.", "iMac20,2") },
+	{ }
+};
+
 static int __init intel_iommu_setup(char *str)
 {
+	const struct dmi_system_id *dmi_id;
+
+	dmi_id = dmi_first_match(force_iommu_quirk);
+	if (dmi_id) {
+		dmar_disabled = 0;
+		pr_info("IOMMU force enabled\n");
+	}
+
 	if (!str)
 		return -EINVAL;
 
-- 
2.25.1

